File: backend/app/schemas/corporate_event_review.py
from datetime import date, datetime
from decimal import Decimal
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class CorporateEventReviewDecision(str, Enum):
    APPROVE = "APPROVE"
    REJECT = "REJECT"


class CorporateEventReviewRequest(BaseModel):
    decision: CorporateEventReviewDecision
    note: str = Field(min_length=10, max_length=2000)


class CorporateExchangeProjectionRequest(BaseModel):
    source_quantity: Decimal = Field(ge=0)
    total_cost: Decimal = Field(ge=0)


class CorporateExchangeProjectionResponse(BaseModel):
    event_id: int
    source_asset_id: int
    destination_asset_id: int
    source_quantity_before: Decimal
    source_quantity_after: Decimal
    destination_quantity_delta: Decimal
    destination_fractional_quantity: Decimal
    total_cost_before: Decimal
    allocated_source_cost: Decimal | None
    allocated_destination_cost: Decimal | None
    cash_component_total: Decimal
    cash_treatment: str | None
    executable: bool
    blocking_reasons: list[str]


class CorporateEventReviewItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    asset_id: int
    ticker: str
    event_type: str
    effective_date: date
    quantity_factor: Decimal
    source_provider: str
    source_event_id: str | None
    reconciliation_status: str
    status: str
    is_canonical: bool
    requires_review: bool
    review_reason: str | None
    reviewed_at: datetime | None
    reviewed_by_user_id: int | None
    review_note: str | None


class CorporateEventReviewPage(BaseModel):
    items: list[CorporateEventReviewItem]
    total: int
    page: int
    page_size: int


class CorporateEventEvidence(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    source_provider: str
    source_event_id: str | None
    event_type: str
    effective_date: date
    record_date: date | None
    ex_date: date | None
    payment_date: date | None
    quantity_factor: Decimal
    cash_component: Decimal | None
    subscription_price: Decimal | None
    destination_cost_allocation: Decimal | None
    quantity_step: Decimal | None
    fractional_settlement_price: Decimal | None
    cash_treatment: str | None
    currency: str
    isin_code: str | None
    destination_isin_code: str | None
    reconciliation_status: str
    status: str
    is_canonical: bool
    raw_metadata: dict[str, object] | None


class CorporateEventEvidenceComparison(BaseModel):
    field: str
    values: dict[str, str | None]
    divergent: bool


class CorporateEventEvidenceGroup(BaseModel):
    selected_event_id: int
    reconciliation_group_hash: str | None
    evidences: list[CorporateEventEvidence]
    comparisons: list[CorporateEventEvidenceComparison]
    economic_effect: str
    terms_complete: bool
    automatic_application_supported: bool
    missing_terms: list[str]
    destination_resolution_status: str | None
    destination_asset_id: int | None
    destination_ticker: str | None
    destination_candidate_ids: list[int]

File: frontend/src/types/corporateEventReview.ts
export type CorporateEventReviewDecision = 'APPROVE' | 'REJECT'

export interface CorporateEventReviewItem {
  id: number
  asset_id: number
  ticker: string
  event_type: string
  effective_date: string
  quantity_factor: string
  source_provider: string
  source_event_id: string | null
  reconciliation_status: string
  status: string
  is_canonical: boolean
  requires_review: boolean
  review_reason: string | null
  reviewed_at: string | null
  reviewed_by_user_id: number | null
  review_note: string | null
}

export interface CorporateEventReviewPage {
  items: CorporateEventReviewItem[]
  total: number
  page: number
  page_size: number
}

export interface CorporateEventReviewFilters {
  page: number
  page_size: number
  ticker?: string
  reconciliation_status?: string
}

export interface CorporateEventEvidence {
  id: number
  source_provider: string
  source_event_id: string | null
  event_type: string
  effective_date: string
  record_date: string | null
  ex_date: string | null
  payment_date: string | null
  quantity_factor: string
  cash_component: string | null
  subscription_price: string | null
  destination_cost_allocation: string | null
  quantity_step: string | null
  fractional_settlement_price: string | null
  cash_treatment: string | null
  currency: string
  isin_code: string | null
  destination_isin_code: string | null
  reconciliation_status: string
  status: string
  is_canonical: boolean
  raw_metadata: Record<string, unknown> | null
}

export interface CorporateEventEvidenceComparison {
  field: string
  values: Record<string, string | null>
  divergent: boolean
}

export interface CorporateEventEvidenceGroup {
  selected_event_id: number
  reconciliation_group_hash: string | null
  evidences: CorporateEventEvidence[]
  comparisons: CorporateEventEvidenceComparison[]
  economic_effect: string
  terms_complete: boolean
  automatic_application_supported: boolean
  missing_terms: string[]
  destination_resolution_status: string | null
  destination_asset_id: number | null
  destination_ticker: string | null
  destination_candidate_ids: number[]
}

export interface CorporateExchangeProjectionPlan {
  event_id: number
  source_asset_id: number
  destination_asset_id: number
  source_quantity_before: string
  source_quantity_after: string
  destination_quantity_delta: string
  destination_fractional_quantity: string
  total_cost_before: string
  allocated_source_cost: string | null
  allocated_destination_cost: string | null
  cash_component_total: string
  cash_treatment: string | null
  executable: boolean
  blocking_reasons: string[]
}
